proposals/data-classes.md

```diff
@@ -0,0 +1,455 @@
+    public string Password { get; set; }
+    public bool RememberMe { get; set; } = false;
+
+    public override bool Equals(object obj)
+        => obj is LoginResource resource && Equals(resource);
+
+    public bool Equals(LoginResource other)
+    {
+        return other != null &&
+               Username == other.Username &&
+               Password == other.Password &&
+               RememberMe == other.RememberMe;
+    }
+
+    public override int GetHashCode()
+    {
+        var hashCode = -736459255;
+        hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Username);
+        hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(Password);
+        hashCode = hashCode * -1521134295 + RememberMe.GetHashCode();
+        return hashCode;
+    }
+
+    public override string ToString()
+    {
+        return $"{{{nameof(Username)} = {Username}, {nameof(Password)} = {Password}, {nameof(RememberMe)} = {RememberMe}}}";
+    }
+
+    public static bool operator ==(LoginResource resource1, LoginResource resource2)
+    {
+        return EqualityComparer<LoginResource>.Default.Equals(resource1, resource2);
+    }
+
+    public static bool operator !=(LoginResource resource1, LoginResource resource2)
+    {
+        return !(resource1 == resource2);
+    }
+}
+```
+
+Immutable data is also a problem. The object initializer syntax provides
+a simple name-based mechanism to create a data type. With `readonly`
+fields or properties, a constructor must be used instead. This creates
+another set of problems:
+
+1. A constructor must be manually defined.
+1. The constructor parameters are ordered, while the fields are not.
+   Consumers have now taken a dependency on the parameter ordering.
+1. The constructor must be maintained with any field changes.
+1. Constructor parameter names don't necessarily line up with field/property
+names hence named argument passing doesn't have the same ease of use as
+object initializers.
+
+There is also no way to create a copy of a data structure with readonly
+fields with only one item changed. A new type must be constructed manually.
+
+## Proposal
+
+To resolve many of these issues, I propose a new modifier for classes and structs: `data`.
+Since this proposal focuses on unorderd, *named* data, I'll refer to this feature as 
+*nominal records*, in contrast to the existing [records proposal](https://github.com/dotnet/csharplang/blob/master/proposals/records.md), which I'll
+refer to as *positional records*. Nominal records are meant to satisfy the
+goals listed above by doing the following things:
+
+1. Automatically generating `Equals`, `GetHashCode`, `ToString`, `==`, `!=`, and `IEquatable<T>`
+   based on the member data of the type.
+1. Allow object initializers to also initialize readonly members.
+
+The LoginResource class now could be defined as
+
+```C#
+public data class LoginResource
+{
+    public string Username { get; }
```

Comment:
Does it even make sense to make these non-public? Could it default to `public`?

Reply:
Private fields and properties are perfectly fine. We could consider it, but right now I'm not wild about changing the default accessibility of members. That's a significant amount of extra complexity in the language, and peoples style guidelines often depend on it.

Reply:
also related, I think it's actually useful to force properties to come first in the class body, otherwise it doesnt look like a "data class" anymore. sure it might be basically a style preference but it worth to consider.